test/format/snap.c

```diff
@@ -492,3 +649,205 @@ snap_repeat_single(WT_CURSOR *cursor, TINFO *tinfo)
+     */
+    newest_oldest_ts = 0ULL;
+    sorted_end = sorted_snaps = dcalloc(SNAP_LIST_SIZE * tinfo_count, sizeof(SNAP_OPS *));
+
+    for (i = 0, tinfop = tinfo_array; i < tinfo_count; ++i, ++tinfop) {
+        tinfo = *tinfop;
+        /*
+         * If this thread has knowledge of the current stable timestamp, that means its "other" snap
+         * list stores up to the stable timestamp, it's the one we want to use. If this thread
+         * doesn't yet have knowledge of the current stable, that means the current snap list is the
+         * one we want.
+         */
+        if (tinfo->stable_ts != g.stable_timestamp)
+            state = tinfo->s;
+        else if (tinfo->s == &tinfo->snap_states[0])
+            state = &tinfo->snap_states[1];
+        else
+            state = &tinfo->snap_states[0];
+        oldest_ts = UINT64_MAX;
+        for (snap = state->snap_state_list; snap < state->snap_state_end; ++snap) {
+            /*
+             * Only keep entries that aren't cleared out and may have relevant timestamps. We don't
+             * fully know which timestamps are relevant, since we haven't computed the newest oldest
+             * yet. We do keep entries that are not marked repeatable, we won't retry unrepeatable
+             * reads, but we need them to invalidate keys that we shouldn't check.
+             */
+            if (snap->op != 0 && snap->ts != 0 && snap->ts <= g.stable_timestamp) {
+                oldest_ts = WT_MIN(oldest_ts, snap->ts);
+                *sorted_end++ = snap;
+            }
+        }
+
+        /*
+         * If there aren't any entries older than the stable timestamp, we've wrapped around. This
+         * thread may have made changes to any key right up to the stable time that have now been
+         * overwritten. There's no way to get an accurate accounting, so we skip checking for this
+         * run.
+         */
+        if (oldest_ts == UINT64_MAX)
+            goto cleanup;
+        newest_oldest_ts = WT_MAX(newest_oldest_ts, oldest_ts);
+    }
+
+    __wt_qsort(
+      sorted_snaps, (size_t)(sorted_end - sorted_snaps), sizeof(SNAP_OPS *), compare_snap_ts);
+
+    /*
+     * Start a new transaction. Verify all repeatable records. Discard the transaction.
+     */
+    while ((ret = session->begin_transaction(session, "isolation=snapshot")) == WT_CACHE_FULL)
+        __wt_yield();
+    testutil_check(ret);
+
+    if (g.c_assert_read_timestamp) {
+        testutil_check(
+          __wt_snprintf(buf, sizeof(buf), "read_timestamp=%" PRIx64, g.stable_timestamp));
+        testutil_check(session->timestamp_transaction(session, buf));
+    }
+
+    testutil_check(session->create(session, "table:wt_snap_keys", "key_format=u,value_format=u"));
+    testutil_check(session->truncate(session, "table:wt_snap_keys", NULL, NULL, NULL));
+    testutil_check(session->open_cursor(session, "table:wt_snap_keys", NULL, NULL, &seen_cursor));
```

Comment:
This scares me because we're doing stuff in the underlying database? Since we're sorting the list, couldn't we sort by name, too, and then it's easy to detect if we've seen the name before because they'll all be in order?

Reply:
Fixed, this all got removed.